File: docs/beginner/module-02-basic-search/code-samples/python/08_search_patterns.py
"""
Search Patterns - Module 2 Python Examples
Common search patterns and strategies in Azure AI Search

This module demonstrates:
- Progressive search strategies
- Search with fallback
- Multi-strategy search
- Search pattern best practices
- When to use different patterns
"""

import os
import sys
import logging
from typing import List, Dict, Any, Optional
from azure.search.documents import SearchClient

# Add setup directory to path for imports
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..', '..', 'setup'))
from connection_utils import get_default_search_client


class SearchPatterns:
    """Class demonstrating common search patterns"""
    
    def __init__(self, search_client: SearchClient = None):
        self.search_client = search_client or get_default_search_client()
        self.logger = logging.getLogger("search_patterns")
    
    def progressive_search(self, query: str, top: int = 10) -> Dict[str, List[Dict[str, Any]]]:
        """
        Progressive search from specific to broad
        
        Args:
            query: Base search query
            top: Maximum results per strategy
            
        Returns:
            Dictionary mapping strategy names to results
        """
        strategies = {}
        
        # 1. Exact phrase (most specific)
        try:
            exact_results = self.search_client.search(
                search_text=f'"{query}"',
                top=top
            )
            strategies['exact_phrase'] = list(exact_results)
            self.logger.info(f"Exact phrase: {len(strategies['exact_phrase'])} results")
        except Exception as e:
            self.logger.error(f"Exact phrase search failed: {str(e)}")
            strategies['exact_phrase'] = []
        
        # 2. All terms (moderate specificity)
        try:
            all_terms_results = self.search_client.search(
                search_text=query,
                search_mode="all",
                top=top
            )
            strategies['all_terms'] = list(all_terms_results)
            self.logger.info(f"All terms: {len(strategies['all_terms'])} results")
        except Exception as e:
            self.logger.error(f"All terms search failed: {str(e)}")
            strategies['all_terms'] = []
        
        # 3. Any terms (broad)
        try:
            any_terms_results = self.search_client.search(
                search_text=query,
                search_mode="any",
                top=top
            )
            strategies['any_terms'] = list(any_terms_results)
            self.logger.info(f"Any terms: {len(strategies['any_terms'])} results")
        except Exception as e:
            self.logger.error(f"Any terms search failed: {str(e)}")
            strategies['any_terms'] = []
        
        # 4. Wildcard (broadest)
        try:
            terms = query.split()
            wildcard_query = " OR ".join([f"{term}*" for term in terms])
            wildcard_results = self.search_client.search(
                search_text=wildcard_query,
                top=top
            )
            strategies['wildcard'] = list(wildcard_results)
            self.logger.info(f"Wildcard: {len(strategies['wildcard'])} results")
        except Exception as e:
            self.logger.error(f"Wildcard search failed: {str(e)}")
            strategies['wildcard'] = []
        
        return strategies
    
    def search_with_fallback(self, query: str, top: int = 10) -> List[Dict[str, Any]]:
        """
        Search with automatic fallback to broader strategies
        
        Args:
            query: Search query
            top: Maximum results to return
            
        Returns:
            List of search results from first successful strategy
        """
        # Try strategies in order of specificity
        strategies = [
            (f'"{query}"', "exact phrase"),
            (query, "all terms (default)"),
            (query, "any terms"),
            (" OR ".join([f"{term}*" for term in query.split()]), "wildcard")
        ]
        
        for search_query, strategy_name in strategies:
            try:
                if strategy_name == "any terms":
                    results = self.search_client.search(
                        search_text=search_query,
                        search_mode="any",
                        top=top
                    )
                else:
                    results = self.search_client.search(
                        search_text=search_query,
                        top=top
                    )
                
                result_list = list(results)
                
                if result_list:
                    self.logger.info(f"Found {len(result_list)} results using {strategy_name}")
                    return result_list
                else:
                    self.logger.info(f"No results with {strategy_name}, trying next strategy")
                    
            except Exception as e:
                self.logger.error(f"Error with {strategy_name}: {str(e)}")
                continue
        
        self.logger.warning("No results found with any search strategy")
        return []
    
    def multi_field_search(self, query: str, field_priority: List[str], top: int = 10) -> List[Dict[str, Any]]:
        """
        Search across fields in order of priority
        
        Args:
            query: Search query
            field_priority: List of fields in priority order
            top: Maximum results to return
            
        Returns:
            Combined results from all fields
        """
        all_results = []
        seen_ids = set()
        
        for field in field_priority:
            try:
                field_results = self.search_client.search(
                    search_text=query,
                    search_fields=[field],
                    top=top
                )
                
                for result in field_results:
                    result_dict = dict(result)
                    result_id = result_dict.get('id', str(hash(str(result_dict))))
                    
                    if result_id not in seen_ids:
                        seen_ids.add(result_id)
                        all_results.append(result_dict)
                
                self.logger.info(f"Field '{field}': found {len(list(field_results))} results")
                
            except Exception as e:
                self.logger.error(f"Error searching field '{field}': {str(e)}")
                continue
        
        # Sort by score
        all_results.sort(key=lambda x: x.get('@search.score', 0), reverse=True)
        return all_results[:top]
    
    def display_progressive_results(self, query: str, strategies: Dict[str, List[Dict[str, Any]]]):
        """Display results from progressive search"""
        print(f"\n🔄 Progressive Search Results: '{query}'")
        print("=" * 60)
        
        strategy_info = {
            'exact_phrase': ('Exact Phrase', 'Most specific - exact phrase match'),
            'all_terms': ('All Terms', 'Moderate - all terms must be present'),
            'any_terms': ('Any Terms', 'Broad - any terms can be present'),
            'wildcard': ('Wildcard', 'Broadest - partial term matching')
        }
        
        for strategy_name, results in strategies.items():
            if strategy_name in strategy_info:
                display_name, description = strategy_info[strategy_name]
                
                print(f"\n{display_name}:")
                print(f"   Description: {description}")
                print(f"   Results: {len(results)} found")
                
                if results:
                    print("   Top matches:")
                    for i, result in enumerate(results[:3], 1):
                        title = result.get('title', 'No title')
                        score = result.get('@search.score', 0.0)
                        print(f"     {i}. {title} (Score: {score:.3f})")
                else:
                    print("   No matches found")
                
                print("-" * 40)
        
        # Recommendation
        print(f"\n💡 RECOMMENDATION:")
        for strategy_name, results in strategies.items():
            if results:
                strategy_display = strategy_info.get(strategy_name, (strategy_name, ''))[0]
                print(f"   Use '{strategy_display}' - found {len(results)} relevant results")
                break
        else:
            print("   Try different search terms or check your data")


def demonstrate_search_patterns():
    """Demonstrate search patterns"""
    print("🎯 Search Patterns Demonstration")
    print("=" * 50)
    
    try:
        # Initialize search patterns
        patterns = SearchPatterns()
        
        # Example 1: Progressive search
        print("\n1️⃣ Progressive Search Strategy")
        
        query = "machine learning"
        progressive_results = patterns.progressive_search(query, top=5)
        patterns.display_progressive_results(query, progressive_results)
        
        # Example 2: Search with fallback
        print(f"\n{'='*70}")
        print("\n2️⃣ Search with Automatic Fallback")
        print("-" * 40)
        
        fallback_query = "artificial intelligence tutorial"
        fallback_results = patterns.search_with_fallback(fallback_query, top=5)
        
        print(f"Query: '{fallback_query}'")
        print(f"Results with fallback: {len(fallback_results)} found")
        
        if fallback_results:
            print("Top results:")
            for i, result in enumerate(fallback_results[:3], 1):
                title = result.get('title', 'No title')
                score = result.get('@search.score', 0.0)
                print(f"  {i}. {title} (Score: {score:.3f})")
        
        # Example 3: Multi-field search
        print(f"\n{'='*70}")
        print("\n3️⃣ Multi-Field Priority Search")
        print("-" * 40)
        
        field_priority = ['title', 'description', 'content', 'tags']
        multi_field_query = "python"
        
        multi_results = patterns.multi_field_search(multi_field_query, field_priority, top=5)
        
        print(f"Query: '{multi_field_query}'")
        print(f"Field priority: {' > '.join(field_priority)}")
        print(f"Combined results: {len(multi_results)} found")
        
        if multi_results:
            print("Top combined results:")
            for i, result in enumerate(multi_results[:3], 1):
                title = result.get('title', 'No title')
                score = result.get('@search.score', 0.0)
                print(f"  {i}. {title} (Score: {score:.3f})")
        
        print("\n✅ Search patterns demonstration completed!")
        
    except Exception as e:
        print(f"❌ Demo failed: {str(e)}")


def search_pattern_best_practices():
    """Best practices for search patterns"""
    print("\n📚 Search Pattern Best Practices")
    print("=" * 50)
    
    print("\n💡 When to Use Each Pattern:")
    
    print("\n🎯 Progressive Search:")
    print("   ✅ When you want comprehensive coverage")
    print("   ✅ For user-facing search interfaces")
    print("   ✅ When result quality is more important than speed")
    print("   ✅ For exploratory or research searches")
    
    print("\n🔄 Fallback Search:")
    print("   ✅ When you need guaranteed results")
    print("   ✅ For automated systems")
    print("   ✅ When speed is important")
    print("   ✅ For simple search interfaces")
    
    print("\n🏗️ Multi-Field Search:")
    print("   ✅ When different fields have different importance")
    print("   ✅ For structured data with clear field hierarchy")
    print("   ✅ When you want to avoid duplicates")
    print("   ✅ For content with rich metadata")
    
    print("\n⚠️ Pattern Selection Guidelines:")
    print("   🔍 Start simple, add complexity as needed")
    print("   📊 Monitor which patterns work best for your data")
    print("   ⚡ Consider performance implications")
    print("   👥 Think about user expectations")
    
    print("\n🔧 Implementation Tips:")
    print("   ✅ Cache results from expensive pattern searches")
    print("   ✅ Log which patterns are most successful")
    print("   ✅ Allow users to choose search modes")
    print("   ✅ Provide feedback about search strategy used")


if __name__ == "__main__":
    # Setup logging
    logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
    
    # Run demonstrations
    demonstrate_search_patterns()
    search_pattern_best_practices()
    
    print("\n💡 Next Steps:")
    print("   - Try different search patterns with your data")
    print("   - Experiment with combining patterns")
    print("   - Consider which patterns work best for your use case")
    print("   - Review all Python examples to build complete search functionality")
    print("   - Check out other language examples in ../csharp/, ../javascript/, etc.")

File: docs/beginner/module-02-basic-search/code-samples/javascript/08_search_patterns.js
/**
 * Search Patterns - Module 2 JavaScript Examples
 * Common search patterns and strategies in Azure AI Search
 * 
 * This module demonstrates:
 * - Progressive search strategies
 * - Search with fallback
 * - Multi-strategy search
 * - Search pattern best practices
 * - When to use different patterns
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');

/**
 * Class demonstrating common search patterns
 */
class SearchPatterns {
    /**
     * Initialize with a search client
     * @param {SearchClient} searchClient - Azure Search client
     */
    constructor(searchClient) {
        if (!searchClient) {
            throw new Error('SearchClient is required');
        }
        this.searchClient = searchClient;
    }

    /**
     * Progressive search from specific to broad
     * @param {string} query - Base search query
     * @param {number} top - Maximum results per strategy
     * @returns {Promise<Object>} Object mapping strategy names to results
     */
    async progressiveSearch(query, top = 10) {
        const strategies = {};

        // 1. Exact phrase (most specific)
        try {
            const exactOptions = { top: top };
            const exactResults = await this.searchClient.search(`"${query}"`, exactOptions);
            
            const exactArray = [];
            for await (const result of exactResults.results) {
                exactArray.push(result);
            }
            
            strategies.exact_phrase = exactArray;
            console.log(`Exact phrase: ${exactArray.length} results`);
        } catch (error) {
            console.error(`Exact phrase search failed: ${error.message}`);
            strategies.exact_phrase = [];
        }

        // 2. All terms (moderate specificity)
        try {
            const allTermsOptions = { 
                top: top,
                searchMode: 'all'
            };
            const allTermsResults = await this.searchClient.search(query, allTermsOptions);
            
            const allTermsArray = [];
            for await (const result of allTermsResults.results) {
                allTermsArray.push(result);
            }
            
            strategies.all_terms = allTermsArray;
            console.log(`All terms: ${allTermsArray.length} results`);
        } catch (error) {
            console.error(`All terms search failed: ${error.message}`);
            strategies.all_terms = [];
        }

        // 3. Any terms (broad)
        try {
            const anyTermsOptions = { 
                top: top,
                searchMode: 'any'
            };
            const anyTermsResults = await this.searchClient.search(query, anyTermsOptions);
            
            const anyTermsArray = [];
            for await (const result of anyTermsResults.results) {
                anyTermsArray.push(result);
            }
            
            strategies.any_terms = anyTermsArray;
            console.log(`Any terms: ${anyTermsArray.length} results`);
        } catch (error) {
            console.error(`Any terms search failed: ${error.message}`);
            strategies.any_terms = [];
        }

        // 4. Wildcard (broadest)
        try {
            const terms = query.split(/\s+/).filter(term => term.length > 0);
            const wildcardQuery = terms.map(term => `${term}*`).join(' OR ');
            
            const wildcardOptions = { top: top };
            const wildcardResults = await this.searchClient.search(wildcardQuery, wildcardOptions);
            
            const wildcardArray = [];
            for await (const result of wildcardResults.results) {
                wildcardArray.push(result);
            }
            
            strategies.wildcard = wildcardArray;
            console.log(`Wildcard: ${wildcardArray.length} results`);
        } catch (error) {
            console.error(`Wildcard search failed: ${error.message}`);
            strategies.wildcard = [];
        }

        return strategies;
    }

    /**
     * Search with automatic fallback to broader strategies
     * @param {string} query - Search query
     * @param {number} top - Maximum results to return
     * @returns {Promise<Array>} Array of search results from first successful strategy
     */
    async searchWithFallback(query, top = 10) {
        // Try strategies in order of specificity
        const strategies = [
            { 
                query: `"${query}"`, 
                name: 'exact phrase',
                options: { top: top }
            },
            { 
                query: query, 
                name: 'all terms (default)',
                options: { top: top }
            },
            { 
                query: query, 
                name: 'any terms',
                options: { top: top, searchMode: 'any' }
            },
            { 
                query: query.split(/\s+/).filter(term => term.length > 0).map(term => `${term}*`).join(' OR '), 
                name: 'wildcard',
                options: { top: top }
            }
        ];

        for (const strategy of strategies) {
            try {
                const results = await this.searchClient.search(strategy.query, strategy.options);
                const resultArray = [];
                
                for await (const result of results.results) {
                    resultArray.push(result);
                }

                if (resultArray.length > 0) {
                    console.log(`Found ${resultArray.length} results using ${strategy.name}`);
                    return resultArray;
                } else {
                    console.log(`No results with ${strategy.name}, trying next strategy`);
                }
            } catch (error) {
                console.error(`Error with ${strategy.name}: ${error.message}`);
                continue;
            }
        }

        console.log('No results found with any search strategy');
        return [];
    }

    /**
     * Search across fields in order of priority
     * @param {string} query - Search query
     * @param {Array<string>} fieldPriority - Array of fields in priority order
     * @param {number} top - Maximum results to return
     * @returns {Promise<Array>} Combined results from all fields
     */
    async multiFieldSearch(query, fieldPriority, top = 10) {
        const allResults = [];
        const seenIds = new Set();

        for (const field of fieldPriority) {
            try {
                const fieldOptions = { 
                    top: top,
                    searchFields: [field]
                };
                
                const fieldResults = await this.searchClient.search(query, fieldOptions);
                let fieldCount = 0;

                for await (const result of fieldResults.results) {
                    const resultId = result.document.id || JSON.stringify(result.document);

                    if (!seenIds.has(resultId)) {
                        seenIds.add(resultId);
                        allResults.push(result);
                    }
                    fieldCount++;
                }

                console.log(`Field '${field}': found ${fieldCount} results`);
            } catch (error) {
                console.error(`Error searching field '${field}': ${error.message}`);
                continue;
            }
        }

        // Sort by score and limit results
        allResults.sort((a, b) => (b.score || 0) - (a.score || 0));
        return allResults.slice(0, top);
    }

    /**
     * Display results from progressive search
     * @param {string} query - Original query
     * @param {Object} strategies - Results from different strategies
     */
    displayProgressiveResults(query, strategies) {
        console.log(`\n🔄 Progressive Search Results: '${query}'`);
        console.log('='.repeat(60));

        const strategyInfo = {
            exact_phrase: {
                displayName: 'Exact Phrase',
                description: 'Most specific - exact phrase match'
            },
            all_terms: {
                displayName: 'All Terms',
                description: 'Moderate - all terms must be present'
            },
            any_terms: {
                displayName: 'Any Terms',
                description: 'Broad - any terms can be present'
            },
            wildcard: {
                displayName: 'Wildcard',
                description: 'Broadest - partial term matching'
            }
        };

        for (const [strategyName, results] of Object.entries(strategies)) {
            const info = strategyInfo[strategyName];
            if (info) {
                console.log(`\n${info.displayName}:`);
                console.log(`   Description: ${info.description}`);
                console.log(`   Results: ${results.length} found`);

                if (results.length > 0) {
                    console.log('   Top matches:');
                    results.slice(0, 3).forEach((result, index) => {
                        const title = result.document.title || 'No title';
                        const score = result.score || 0.0;
                        console.log(`     ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
                    });
                } else {
                    console.log('   No matches found');
                }

                console.log('-'.repeat(40));
            }
        }

        // Recommendation
        console.log('\n💡 RECOMMENDATION:');
        for (const [strategyName, results] of Object.entries(strategies)) {
            if (results.length > 0) {
                const info = strategyInfo[strategyName];
                if (info) {
                    console.log(`   Use '${info.displayName}' - found ${results.length} relevant results`);
                    break;
                }
            }
        }

        const hasResults = Object.values(strategies).some(results => results.length > 0);
        if (!hasResults) {
            console.log('   Try different search terms or check your data');
        }
    }
}

/**
 * Demonstrate search patterns
 */
async function demonstrateSearchPatterns() {
    console.log('🎯 Search Patterns Demonstration');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const patterns = new SearchPatterns(searchClient);

        // Example 1: Progressive search
        console.log('\n1️⃣ Progressive Search Strategy');

        const query = 'machine learning';
        const progressiveResults = await patterns.progressiveSearch(query, 5);
        patterns.displayProgressiveResults(query, progressiveResults);

        // Example 2: Search with fallback
        console.log('\n' + '='.repeat(70));
        console.log('\n2️⃣ Search with Automatic Fallback');
        console.log('-'.repeat(40));

        const fallbackQuery = 'artificial intelligence tutorial';
        const fallbackResults = await patterns.searchWithFallback(fallbackQuery, 5);

        console.log(`Query: '${fallbackQuery}'`);
        console.log(`Results with fallback: ${fallbackResults.length} found`);

        if (fallbackResults.length > 0) {
            console.log('Top results:');
            fallbackResults.slice(0, 3).forEach((result, index) => {
                const title = result.document.title || 'No title';
                const score = result.score || 0.0;
                console.log(`  ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
            });
        }

        // Example 3: Multi-field search
        console.log('\n' + '='.repeat(70));
        console.log('\n3️⃣ Multi-Field Priority Search');
        console.log('-'.repeat(40));

        const fieldPriority = ['title', 'description', 'content', 'tags'];
        const multiFieldQuery = 'python';

        const multiResults = await patterns.multiFieldSearch(multiFieldQuery, fieldPriority, 5);

        console.log(`Query: '${multiFieldQuery}'`);
        console.log(`Field priority: ${fieldPriority.join(' > ')}`);
        console.log(`Combined results: ${multiResults.length} found`);

        if (multiResults.length > 0) {
            console.log('Top combined results:');
            multiResults.slice(0, 3).forEach((result, index) => {
                const title = result.document.title || 'No title';
                const score = result.score || 0.0;
                console.log(`  ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
            });
        }

        console.log('\n✅ Search patterns demonstration completed!');

    } catch (error) {
        console.error(`❌ Demo failed: ${error.message}`);
    }
}

/**
 * Display search pattern best practices
 */
function searchPatternBestPractices() {
    console.log('\n📚 Search Pattern Best Practices');
    console.log('='.repeat(50));

    console.log('\n💡 When to Use Each Pattern:');

    console.log('\n🎯 Progressive Search:');
    console.log('   ✅ When you want comprehensive coverage');
    console.log('   ✅ For user-facing search interfaces');
    console.log('   ✅ When result quality is more important than speed');
    console.log('   ✅ For exploratory or research searches');

    console.log('\n🔄 Fallback Search:');
    console.log('   ✅ When you need guaranteed results');
    console.log('   ✅ For automated systems');
    console.log('   ✅ When speed is important');
    console.log('   ✅ For simple search interfaces');

    console.log('\n🏗️ Multi-Field Search:');
    console.log('   ✅ When different fields have different importance');
    console.log('   ✅ For structured data with clear field hierarchy');
    console.log('   ✅ When you want to avoid duplicates');
    console.log('   ✅ For content with rich metadata');

    console.log('\n⚠️ Pattern Selection Guidelines:');
    console.log('   🔍 Start simple, add complexity as needed');
    console.log('   📊 Monitor which patterns work best for your data');
    console.log('   ⚡ Consider performance implications');
    console.log('   👥 Think about user expectations');

    console.log('\n🔧 Implementation Tips:');
    console.log('   ✅ Cache results from expensive pattern searches');
    console.log('   ✅ Log which patterns are most successful');
    console.log('   ✅ Allow users to choose search modes');
    console.log('   ✅ Provide feedback about search strategy used');

    console.log('\n🚀 Performance Considerations:');
    console.log('   ✅ Use async/await for better performance');
    console.log('   ✅ Implement proper error handling');
    console.log('   ✅ Consider result caching strategies');
    console.log('   ✅ Monitor search latency and adjust patterns');
}

// Main execution
async function main() {
    try {
        await demonstrateSearchPatterns();
        searchPatternBestPractices();

        console.log('\n💡 Next Steps:');
        console.log('   - Try different search patterns with your data');
        console.log('   - Experiment with combining patterns');
        console.log('   - Consider which patterns work best for your use case');
        console.log('   - Review all JavaScript examples to build complete search functionality');
        console.log('   - Check out other language examples in ../python/, ../csharp/, etc.');
    } catch (error) {
        console.error(`❌ Main execution failed: ${error.message}`);
    }
}

// Export for use in other modules
module.exports = {
    SearchPatterns,
    demonstrateSearchPatterns,
    searchPatternBestPractices
};

// Run if this file is executed directly
if (require.main === module) {
    main();
}